File: projects/17-MinkowskiSum/implementation/minkowskiSum.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static List<Point> input;
    public static void main(String[] args){
        int n = sc.nextInt(); // n = number of vertices in the first polygon
        int m = sc.nextInt(); // m = number of vertices in the second polygon
        Polygon A = parsePolygon(n); // takes next n lines to be the points of Polygon A
        Polygon B = parsePolygon(m); // takes next m lines to be the points of Polygon B
        Polygon SUM = A.add(B); // Minkowski Sum of Convex Polygons
        System.out.println(SUM);
    }

    // parses the next n lines for points, using them to construct a polygon
    public static Polygon parsePolygon(int n){
        input = new ArrayList<>(n);
        for(int i = 0; i < n; i++){
            double x = sc.nextDouble();
            double y = sc.nextDouble();
            input.add(new Point(x, y));
        }
        return new Polygon(input);
    }
}

class Polygon {
    List<Point> points;
    List<Edge> edges;
    public Polygon(List<Point> input){
        constructPoints(input);
        constructEdges();
    }
    // We construct the list of points by starting at one of the lowest points and moving counterclockwise
    public void constructPoints(List<Point> input){
        points = new ArrayList<>(input.size());
        Point root = input.get(0);
        int start = 0;
        // When we find a new lowest y-value, it becomes the new root
        // If we have a tie, choose the one with the lowest x-value
        for(int i = 1; i < input.size(); i++){
            Point p = input.get(i);
            if(p.y < root.y){
                root = p;
                start = i;
            }
            else if(p.y == root.y && p.x < root.x){
                root = p;
                start = i;
            }
        }
        // Cyclically add the roots counterclockwise starting from the root
        for(int i=0; i<input.size(); i++){
            points.add(input.get((i+start)%input.size()));
        }
    }
    // We construct the directed edges by starting at the root and moving counterclockwise
    public void constructEdges(){
        edges = new ArrayList<>(points.size());
        if(points.size() <= 1) return;
        for(int i=0; i<points.size()-1; i++){
            edges.add(new Edge(points.get(i), points.get(i+1)));
        }
        edges.add(new Edge(points.get(points.size()-1), points.get(0)));
    }
    // The Minkowski Sum of Convex Polygons
    public Polygon add(Polygon p){
        // Collect the edges into a list sorted by radial angle from the positive x-axis
        List<Edge> merged = mergeEdges(edges, p.edges);
        List<Point> res = new ArrayList<>();
        // Begin by adding the roots of each polygon. This is our initial vertex.
        Point cur = points.get(0).add(p.points.get(0));
        // Edge case if both inputs are single points.
        if(merged.isEmpty()){
            res.add(cur);
            return new Polygon(res);
        }
        // For every directed edge, move the current vertex in that direction (via vector addition).
        for(Edge e : merged){
            res.add(cur);
            cur = cur.add(e.direction());
        }
        // Use the list of vertices to construct the polygon which is the sum of this and p.
        return new Polygon(res);
    }
    // Merges two sorted list into one list sorted by radial angle from the positive x-axis.
    public static List<Edge> mergeEdges(List<Edge> list1, List<Edge> list2){
        int size = list1.size() + list2.size();
        List<Edge> res = new ArrayList<>(size);
        int p1 = 0;
        int p2 = 0;
        Edge last = null;
        // While there are still edges in both lists left to add,
        // look at the two vectors v1 and v2 which are pointed to
        // by p1 and p2, respectively. If v1 cross v2 is positive,
        // then that indicates a counter-clockwise turn from v1 to
        // v2, so v1 must have a smaller radial angle, and thus goes
        // first. If one list has already been exhausted, simply
        // add the remaining edges from the other list.
        while(p1<list1.size() || p2<list2.size()){
            Edge edge;
            if(p1>=list1.size()){
                edge = list2.get(p2++);
            }
            else if(p2>=list2.size()){
                edge = list1.get(p1++);
            }
            else{
                double orientation = list1.get(p1).orientation(list2.get(p2));
                if(orientation > 0){
                    edge = list1.get(p1++);
                }
                else{
                    edge = list2.get(p2++);
                }
            }
            // This eliminates collinear points from the output.
            // If the previous edge is parallel to the last edge,
            // merge them, remove the previous edge, and add the
            // new merged one.
            if(edge.isInDirectionOf(last)){
                edge = edge.merge(last);
                res.remove(res.size()-1);
            }
            res.add(edge);
            last = edge;
        }
        return res;
    }
    public String toString(){
        StringBuilder sb = new StringBuilder();
        for(Point p : points){
            sb.append(p);
            sb.append("\n");
        }
        return sb.toString();
    }
}

// Represents a directed edge from Point a to Point b
class Edge{
    Point a, b;
    final static Edge AXIS = new Edge(new Point(0,0), new Point(1,0)); // positive x-axis
    public Edge(Point a, Point b){
        this.a = a;
        this.b = b;
    }
    // The vector representation of the edge direction
    public Point direction(){
        return new Point(b.x-a.x, b.y-a.y);
    }
    // First checkes if the directions are parallel, then the dot product ensures they are not in opposite directions.
    public boolean isInDirectionOf(Edge e){
        if(cross(e)!=0) return false;
        return direction().dot(e.direction()) > 0;
    }
    // Vector addition of edges
    public Edge merge(Edge e){
        return new Edge(a, b.add(e.direction()));
    }
    // Cross product of the vector representation of the edges
    public double cross(Edge e){
        if(e==null) return Double.MIN_VALUE;
        double x1 = b.x-a.x;
        double y1 = b.y-a.y;
        double x2 = e.b.x-e.a.x;
        double y2 = e.b.y-e.a.y;
        return x1*y2 - x2*y1;
    }
    public double dot(Edge e){
        return direction().dot(e.direction());
    }
    public int orientation(Edge e){
        double c1 = AXIS.cross(this);
        double c2 = AXIS.cross(e);
        double d1 = AXIS.dot(this);
        double d2 = AXIS.dot(e);
        if(c1==0 && d1 > 0) return 1;
        if(c2==0 && d2 > 0) return -1;
        if(c1 > 0 && c2 < 0) return 1;
        if(c1 < 0 && c2 > 0) return -1;
        double c3 = cross(e);
        if(c3 > 0) return 1;
        return -1;
    }
    public double magnitude(){
        return direction().magnitude();
    }
    public String toString(){
        return "[" + a.toString() + ", " + b.toString() + "]";
    }
}

// A simple pair of coordinates in the cartesian plane
class Point{
    double x, y;
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
    public double dot(Point p){
        return x*p.x + y*p.y;
    }
    // Vector addition of points
    public Point add(Point p){
        return new Point(x+p.x, y+p.y);
    }
    public double magnitude(){
        return Math.sqrt(x*x + y*y);
    }
    public String toString(){
        return x + " " + y;
    }
}

File: projects/17-MinkowskiSum/implementation/minkowskiSum.cpp
#include <vector>
#include <iostream>
using namespace std;

struct point {
    double x;
    double y;

    point(double x, double y) {
        this->x = x;
        this->y = y;
    };
};

// Takes two convex polygons, represented as sets of vertices in counterclockwise order
// and returns their minkowski sum, represented in the same way.
// For this implementation
vector<point> minkowskiSum(vector<point>& polyA, vector<point>& polyB) {
    const int n = polyA.size(), m = polyB.size();
    vector<point> polySum;

    // Indices i and j are used as the starting positions for the algorithm,
    // as we must ensure that the sum of the first two vertices is a vertex
    // in the minkowski sum. Getting the bottom left points of both polygons 
    // accomplishes this.
    int i = 0;
    int j = 0;

    // Get index of bottom left point in A
    point bottomLeftPoint = polyA[0];
    for (int k=1; k < n; k++) {
        if (polyA[k].y < bottomLeftPoint.y || 
                (polyA[k].y == bottomLeftPoint.y && polyA[k].x < bottomLeftPoint.x)) {
            bottomLeftPoint = polyA[k];
            i = k;
        }
    }

    // Get index of bottom left point in B
    bottomLeftPoint = polyB[0];
    for (int k=1; k < m; k++) {
        if (polyB[k].y < bottomLeftPoint.y || 
                (polyB[k].y == bottomLeftPoint.y && polyB[k].x < bottomLeftPoint.x)) {
            bottomLeftPoint = polyB[k];
            j = k;
        }
    }


    int t = 0; // number of points we have moved through in A
    int s = 0; // number of points we have moved through in B
    while (t+s < n+m) {
        // Add new point corresponding to current index positions
        polySum.emplace_back(polyA[i].x + polyB[j].x, polyA[i].y + polyB[j].y);

        // Find indices from a cyclic shift
        int i1 = i+1;
        if (i1 == n) i1 = 0;
        int j1 = j+1;
        if (j1 == m) j1 = 0;

        // By math, stores which polar angle is higher between A[i] -> A[i+1]
        // and B[j] -> B[j+1]. Positive if first polar angle is higher,
        // zero if equivalent polar angles, and negative otherwise.
        // More concretely, this is the cross product of (polyB[j1] - polyB[j]) and (polyA[i1] - polyA[i])
        double crossProd = (polyA[i1].y - polyA[i].y)*(polyB[j1].x - polyB[j].x) - (polyB[j1].y - polyB[j].y) *(polyA[i1].x - polyA[i].x);

        // The cross product tells us which edge has higher polar angle. 
        // If A[i] -> A[i+1] is higher polar angle,
        // move to the next point in B. If they are equal,
        // we move to the next point in both. If A[i] -> A[i+1] is lower,
        // we move to the next point in A. This ensures we are iterating
        // forward by increasing polar angle of the edge segments
        if (crossProd > 0) {
            j = j1;
            s++;
        } else if (crossProd == 0) {
            i = i1;
            j = j1;
            t++; s++;
        } else {
            i = i1;
            t++;
        }
    }
    return polySum;
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<point> a;
    for (int i=0; i < n; i++) {
        double x, y;
        cin >> x >> y;
        a.emplace_back(x, y);
    }
    vector<point> b;
    for (int i=0; i < m; i++) {
        double x, y;
        cin >> x >> y;
        b.emplace_back(x, y);
    }

    vector<point> c = minkowskiSum(a, b);
    for (auto& p : c) {
        cout << p.x << ' ' << p.y << endl;
    }
    cout << endl;
}
